File: hamster/v1.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MAX_N 50
#define MAX_L 200
#define MAX_B 200
#define MAX_S 50
#define MAX_STEPS 2000

// =========================
// STRUCTURES
// =========================

typedef struct {
    int x, y, r;
} Light;

typedef struct {
    int x, y, g, s;
} Block;

typedef struct {
    int x, y, g;
} Switch;

typedef struct {
    int x, y;
} Pos;

// =========================
// GLOBAL GRID
// =========================

int N;

Pos hamster;

Light lights[MAX_L];
Block blocks[MAX_B];
Switch switches[MAX_S];

int l_count = 0;
int b_count = 0;
int s_count = 0;

// =========================
// UTILS
// =========================

int sign(int a) {
    return (a > 0) - (a < 0);
}

int in_bounds(int x, int y) {
    return x >= 0 && y >= 0 && x < N && y < N;
}

// =========================
// VISIBILITE LUMIERE
// =========================

int ray_visible(int x, int y, Light *l) {

    int dx = l->x - x;
    int dy = l->y - y;

    if (dx != 0 && dy != 0) return 0;

    int dist = abs(dx + dy);
    if (dist > l->r) return 0;

    int sx = sign(dx);
    int sy = sign(dy);

    int cx = x + sx;
    int cy = y + sy;

    while (!(cx == l->x && cy == l->y)) {

        // blocs
        for (int i = 0; i < b_count; i++) {
            if (blocks[i].x == cx && blocks[i].y == cy && blocks[i].s == 1)
                return 0;
        }

        cx += sx;
        cy += sy;
    }

    return 1;
}

// =========================
// GET DIRECTION LIGHT
// =========================

int get_direction(int x, int y, int *dx, int *dy) {

    int count = 0;
    int tmp_dx = 0, tmp_dy = 0;

    for (int i = 0; i < l_count; i++) {

        if (ray_visible(x, y, &lights[i])) {

            int vx = lights[i].x - x;
            int vy = lights[i].y - y;

            if (vx != 0) {
                tmp_dx = -sign(vx);
                tmp_dy = 0;
            } else {
                tmp_dx = 0;
                tmp_dy = -sign(vy);
            }

            count++;
        }
    }

    if (count == 1) {
        *dx = tmp_dx;
        *dy = tmp_dy;
    }

    return count;
}

// =========================
// SIMULATION
// =========================
int simulate() {

    int x = hamster.x;
    int y = hamster.y;

    int dx = 0;
    int dy = 0;

    int steps = 0;

    for (int i = 0; i < MAX_STEPS; i++) {

        int vis_count = 0;

        int tmp_dx = 0;
        int tmp_dy = 0;

        // =========================
        // DETECTION LUMIERES
        // =========================
        for (int l = 0; l < l_count; l++) {

            Light *light = &lights[l];

            int vx = light->x - x;
            int vy = light->y - y;

            // pas même ligne/colonne
            if (vx != 0 && vy != 0)
                continue;

            int dist = abs(vx + vy);
            if (dist > light->r)
                continue;

            int sx = sign(vx);
            int sy = sign(vy);

            int cx = x + sx;
            int cy = y + sy;

            int blocked = 0;

            // check obstacles entre hamster et lumière
            while (!(cx == light->x && cy == light->y)) {

                // blocs actifs bloquent vision
                for (int b = 0; b < b_count; b++) {
                    if (blocks[b].x == cx &&
                        blocks[b].y == cy &&
                        blocks[b].s == 1) {
                        blocked = 1;
                        break;
                    }
                }

                if (blocked) break;

                cx += sx;
                cy += sy;
            }

            if (blocked) continue;

            // lumière visible
            vis_count++;

            tmp_dx = (vx != 0) ? -sign(vx) : 0;
            tmp_dy = (vy != 0) ? -sign(vy) : 0;
        }

        // =========================
        // REGLES DE DECISION
        // =========================

        if (vis_count >= 2)
            return steps;

        if (vis_count == 1) {
            dx = tmp_dx;
            dy = tmp_dy;
        }

        if (vis_count == 0) {
            if (dx == 0 && dy == 0)
                return steps;
            // sinon on continue direction actuelle
        }

        // =========================
        // DEPLACEMENT
        // =========================

        int nx = x + dx;
        int ny = y + dy;

        if (nx < 0 || ny < 0 || nx >= N || ny >= N)
            return steps;

        // collision blocs
        for (int b = 0; b < b_count; b++) {
            if (blocks[b].x == nx &&
                blocks[b].y == ny &&
                blocks[b].s == 1)
                return steps;
        }

        // déplacement
        x = nx;
        y = ny;
        steps++;
    }

    return 0;
}	
// =========================
// DESIGN SIMPLE (CORRIDOR)
// =========================
void build_corridor() {

    hamster.x = 1;
    hamster.y = N / 2;

    l_count = 0;
    b_count = 0;

    int y = N / 2;

    // 🔥 LUMIERE DE START OBLIGATOIRE
    lights[l_count++] = (Light){2, y, N};

    // chaîne directionnelle
    for (int x = 3; x < N - 2; x++) {
        if (x % 2 == 0) {
            lights[l_count++] = (Light){x, y, N};
        }
    }

    // murs latéraux (ok)
    for (int i = 0; i < N; i++) {
        blocks[b_count++] = (Block){0, i, 0, 1};
        blocks[b_count++] = (Block){N - 1, i, 0, 1};
    }
}

// =========================
// EXPORT
// =========================

void export_solution(int score) {

    char filename[64];
    sprintf(filename, "%d_c_%d.txt", N, score);

    FILE *f = fopen(filename, "w");

    fprintf(f, "%d\n", N);
    fprintf(f, "(%d, %d)\n", hamster.x, hamster.y);

    for (int i = 0; i < l_count; i++) {
        fprintf(f, "L (%d, %d) %d\n",
                lights[i].x,
                lights[i].y,
                lights[i].r);
    }

    for (int i = 0; i < b_count; i++) {
        fprintf(f, "B (%d, %d) %d %d\n",
                blocks[i].x,
                blocks[i].y,
                blocks[i].g,
                blocks[i].s);
    }

    for (int i = 0; i < s_count; i++) {
        fprintf(f, "S (%d, %d) %d\n",
                switches[i].x,
                switches[i].y,
                switches[i].g);
    }

    fclose(f);
}

// =========================
// MAIN
// =========================

int main() {

    srand(time(NULL));

    N = 10;

    build_corridor();

    int best_score = 0;

    for (int i = 0; i < 1000; i++) {

        int s = simulate();

        if (s > best_score)
            best_score = s;
    }

    export_solution(best_score);

    printf("BEST SCORE = %d\n", best_score);

    return 0;
}

File: rendu/hamster/challenge/v1.py
import copy

# =========================
# CONFIG
# =========================

MAX_STEPS = 5000

# =========================
# STRUCTURES
# =========================

class Light:
    def __init__(self, x, y, r):
        self.x = x
        self.y = y
        self.r = r


class Block:
    def __init__(self, x, y, g, s):
        self.x = x
        self.y = y
        self.g = g
        self.s = s


class Switch:
    def __init__(self, x, y, g):
        self.x = x
        self.y = y
        self.g = g


# =========================
# UTILS
# =========================

def sign(a):
    return (a > 0) - (a < 0)


# =========================
# PARSER
# =========================

def parse_file(filename):

    with open(filename, "r") as f:
        lines = [line.strip() for line in f.readlines() if line.strip()]

    N = int(lines[0])

    # hamster ignoré
    # ligne 1 = "(x, y)"
    # on l'utilise pas

    lights = []
    blocks = []
    switches = []

    occupied = set()

    for line in lines[2:]:

        parts = line.split()

        # =========================
        # LIGHT
        # =========================
        if line.startswith("L"):

            x = int(parts[1][1:-1].split(",")[0])
            y = int(parts[2][:-1])

            r = int(parts[3])

            lights.append(Light(x, y, r))
            occupied.add((x, y))

        # =========================
        # BLOCK
        # =========================
        elif line.startswith("B"):

            x = int(parts[1][1:-1].split(",")[0])
            y = int(parts[2][:-1])

            g = int(parts[3])
            s = int(parts[4])

            blocks.append(Block(x, y, g, s))
            occupied.add((x, y))

        # =========================
        # SWITCH
        # =========================
        elif line.startswith("S"):

            x = int(parts[1][1:-1].split(",")[0])
            y = int(parts[2][:-1])

            g = int(parts[3])

            switches.append(Switch(x, y, g))
            occupied.add((x, y))

    return {
        "N": N,
        "lights": lights,
        "blocks": blocks,
        "switches": switches,
        "occupied": occupied
    }


# =========================
# VISIBILITY
# =========================

def ray_visible(grid, x, y, light):

    dx = light.x - x
    dy = light.y - y

    # pas même ligne/colonne
    if dx != 0 and dy != 0:
        return False

    dist = abs(dx + dy)

    if dist > light.r:
        return False

    step_x = sign(dx)
    step_y = sign(dy)

    cx = x + step_x
    cy = y + step_y

    while (cx, cy) != (light.x, light.y):

        # blocs actifs
        for b in grid["blocks"]:
            if b.x == cx and b.y == cy and b.s == 1:
                return False

        # autres lumières
        for l in grid["lights"]:
            if l.x == cx and l.y == cy:
                return False

        cx += step_x
        cy += step_y

    return True


def get_visible_lights(grid, x, y):

    visible = []

    for l in grid["lights"]:

        if ray_visible(grid, x, y, l):

            dx = l.x - x
            dy = l.y - y

            if dx != 0:
                direction = (-sign(dx), 0)
            else:
                direction = (0, -sign(dy))

            visible.append(direction)

    return visible


# =========================
# SIMULATION
# =========================

def simulate(grid, hamster_pos):

    local_grid = copy.deepcopy(grid)

    N = local_grid["N"]

    x, y = hamster_pos

    dx, dy = 0, 0

    steps = 0

    visited = set()

    for _ in range(MAX_STEPS):

        state = (x, y, dx, dy)

        if state in visited:
            return 0

        visited.add(state)

        visible = get_visible_lights(local_grid, x, y)

        # 2 lumières visibles = stop
        if len(visible) >= 2:
            return steps

        # 1 lumière = direction opposée
        if len(visible) == 1:
            dx, dy = visible[0]

        # aucune lumière
        else:
            if dx == 0 and dy == 0:
                return steps

        nx = x + dx
        ny = y + dy

        # hors grille
        if not (0 <= nx < N and 0 <= ny < N):
            return steps

        # collision blocs actifs
        blocked = False

        for b in local_grid["blocks"]:
            if b.x == nx and b.y == ny and b.s == 1:
                blocked = True
                break

        if blocked:
            return steps

        # collision lumières
        for l in local_grid["lights"]:
            if l.x == nx and l.y == ny:
                blocked = True
                break

        if blocked:
            return steps

        # switches
        for s in local_grid["switches"]:
            if s.x == nx and s.y == ny:

                for b in local_grid["blocks"]:
                    if b.g == s.g:
                        b.s = 1 - b.s

        x = nx
        y = ny

        steps += 1

    return 0


# =========================
# TEST
# =========================

def test(namefile):

    grid = parse_file(namefile)

    N = grid["N"]

    best_score = -1
    best_positions = []

    # toutes les cases libres
    for x in range(N):
        for y in range(N):

            if (x, y) in grid["occupied"]:
                continue

            score = simulate(grid, (x, y))

            print(f"TEST ({x}, {y}) => {score}")

            if score > best_score:
                best_score = score
                best_positions = [(x, y)]

            elif score == best_score:
                best_positions.append((x, y))

    print("\n========================")
    print("BEST SCORE:", best_score)
    print("BEST POSITIONS:")

    for pos in best_positions:
        print(pos)

    return best_positions, best_score


# =========================
# RUN
# =========================

if __name__ == "__main__":

    test("10_challenge3_1.txt")